HiP-ThumbnailService/Controllers/ThumbnailsController.cs

```diff
@@ -0,0 +1,160 @@
+
+        public ThumbnailsController(IOptions<DirConfig> uploadConfig, IOptions<SizeConfig> sizeConfig, IOptions<EndpointConfig> endpointConfig)
+        {
+            _dirConfig = uploadConfig.Value;
+            _sizeConfig = sizeConfig.Value;
+            _endpointConfig = endpointConfig.Value;
+        }
+
+        [ProducesResponseType(204)]
+        [HttpDelete]
+        public async Task<IActionResult> Put(string url)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest(ModelState);
+            
+
+            var encodedId = Convert.ToBase64String(Encoding.UTF8.GetBytes(url));
+            var folderPath = Path.Combine(_dirConfig.Path, encodedId);
+
+            var semaphore = LockDictionary.GetOrAdd(encodedId, new SemaphoreSlim(1));
+
+            await semaphore.WaitAsync();
+            try
+            {
+                //clear folder
+                if (Directory.Exists(folderPath))
+                {
+                    var folder = new DirectoryInfo(folderPath);
+                    folder.Delete(true);
+                }
+                
+                return NoContent();
+            }
+            finally
+            {
+                semaphore.Release();
+            }
+        }
+
+        [ProducesResponseType(400)]
+        [ProducesResponseType(200)]
+        [HttpGet]
+        public async Task<IActionResult> Get(CreationArgs args)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest(ModelState);
+
+            if (string.IsNullOrEmpty(args.Url)) return BadRequest(new { Message = "An url for image access must be provided" });
+
+            var encodedId = Convert.ToBase64String(Encoding.UTF8.GetBytes(args.Url));
+            var folderPath = Path.Combine(_dirConfig.Path, encodedId);
+
+            if (string.IsNullOrEmpty(args.Size) || !_sizeConfig.SupportedSizes.ContainsKey(args.Size))
+                return BadRequest(new { Message = "Invalid size" });
+
+            var semaphore = LockDictionary.GetOrAdd(encodedId, new SemaphoreSlim(1));
+            await semaphore.WaitAsync();
+            try
+            {
+                var extension = RequestImageFormatUtils.GetExtension(args.Format);
+                var thumbnailPath = Path.Combine(folderPath, GetFileName(args.Size, args.Mode)) + "." +
+                                    extension;
+                if (!System.IO.File.Exists(thumbnailPath))
+                {
+                    var client = new HttpClient();
+                    client.DefaultRequestHeaders.Add("Authorization", Request.Headers["Authorization"].ToString());
+                    var hostUrl = _endpointConfig.HostUrl;
+                    var stream = await client.GetStreamAsync(hostUrl + args.Url);
+
+                    if (stream == null)
+                    {
+                        return BadRequest(new { Message = "Image access failed" });
```

Comment:
1. `400 Bad Request` is inappropiate here: "The HTTP 400 Bad Request response status code indicates that the server could not understand the request due to invalid syntax" - [MDN web docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400). How about `404 Not Found`?

1. `GetStreamAsync` never returns null, so this code is never reached and `500 Internal Server Error` is returned if, for example, the specified URL returns 404. You need `try/catch` here.